### PR TITLE
Shared by links added to app drawer [NEW]

### PR DESCRIPTION
**Pull request by davigonz**
Fixes https://github.com/owncloud/android/issues/2276
To merge https://github.com/owncloud/android/pull/2546

Contribution by @sd1998 

QA

Test Plan: https://github.com/owncloud/QA/blob/master/Mobile/Android/Release_2.14.0/2705-%20share_links_drawer.md

- [x] (1) Sort by name is not correct https://github.com/owncloud/android/pull/2705#issuecomment-555400930
- [x] (2) Folders browsability https://github.com/owncloud/android/pull/2705#issuecomment-555408073 [FIXED]
- [x] (3) List with 0B items https://github.com/owncloud/android/pull/2705#issuecomment-561121474 [FIXED]
- [x] (4) improvement: getting shares form API https://github.com/owncloud/android/pull/2705#issuecomment-561125480 [WILL NOT BE FIXED HERE]

### CODE REVIEW

File: owncloudApp/src/main/java/com/owncloud/android/datamodel/FileDataStorageManager.java [hunk 1]
```diff
@@ -2057,4 +2061,47 @@ private RemoteCapability createCapabilityInstance(Cursor c) {
...
     }
+
+    public Vector<OCFile> getSharedByLinkFilesFromCurrentAccount() {
+        Vector<OCFile> allSharedFiles = new Vector<OCFile>();
+        Vector<OCFile> result = new Vector<OCFile>();
+        Cursor cursorOnShared = null;
+        try {
+            cursorOnShared = getContentResolver().query(ProviderTableMeta.CONTENT_URI,
+                    null,
+                     ProviderTableMeta.FILE_SHARED_VIA_LINK + " = ? AND " +
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/datamodel/FileDataStorageManager.java]: This seems to be not properly indented

File: owncloudApp/src/main/java/com/owncloud/android/files/FileMenuFilter.java [hunk 2]
```diff
@@ -91,15 +92,15 @@ public FileMenuFilter(OCFile targetFile, Account account, ComponentsGetter cg,
      * @param menu Options or context menu to filter.
      */
     public void filter(Menu menu, boolean displaySelectAll, boolean displaySelectInverse,
-                       boolean onlyAvailableOffline) {
+                       boolean onlyAvailableOffline,boolean sharedByLinkFiles) {
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/files/FileMenuFilter.java]: Space after `,`

File: owncloudApp/src/main/java/com/owncloud/android/files/FileMenuFilter.java [hunk 3]
```diff
@@ -91,15 +92,15 @@ public FileMenuFilter(OCFile targetFile, Account account, ComponentsGetter cg,
...
         if (mFiles == null || mFiles.size() <= 0) {
             hideAll(menu);
 
         } else {
             List<Integer> toShow = new ArrayList<>();
             List<Integer> toHide = new ArrayList<>();
 
-            filter(toShow, toHide, displaySelectAll, displaySelectInverse, onlyAvailableOffline);
+            filter(toShow, toHide, displaySelectAll, displaySelectInverse, onlyAvailableOffline,sharedByLinkFiles);
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/files/FileMenuFilter.java]: Space before `sharedByLinkFiles`. Please indent this file.

File: owncloudApp/src/main/java/com/owncloud/android/operations/SynchronizeFolderOperation.java [hunk 4]
```diff
@@ -323,7 +324,7 @@ private void mergeRemoteFolder(ArrayList<RemoteFile> remoteFolderAndFiles)
         }
 
         // get current data about local contents of the folder to synchronize
-        List<OCFile> localFiles = storageManager.getFolderContent(mLocalFolder, false);
+        List<OCFile> localFiles = storageManager.getFolderContent(mLocalFolder, false,false);
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/operations/SynchronizeFolderOperation.java]: Indent this file

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt [hunk 5]
```diff
@@ -510,8 +511,7 @@ class DocumentsStorageProvider : DocumentsProvider() {
     private fun findFiles(root: OCFile, query: String): Vector<OCFile> {
         val result = Vector<OCFile>()
 
-        val folderContent = currentStorageManager?.getFolderContent(root, false) ?: return result
-
+        val folderContent = currentStorageManager?.getFolderContent(root, false,false) ?: return result
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: Same

File: owncloudApp/src/main/java/com/owncloud/android/ui/activity/FileActivity.java [hunk 6]
```diff
@@ -473,6 +474,14 @@ public void onlyAvailableOfflineOption() {
         startActivity(intent);
     }
 
+     @Override
+     public void sharedByLinkFilesOption(){
+        Intent intent = new Intent(this,FileDisplayActivity.class);
+        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
+        intent.putExtra(EXTRA_SHARED_BY_LINK_FILES,true);
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/ui/activity/FileActivity.java]: Same

File: owncloudApp/src/main/java/com/owncloud/android/ui/activity/ReceiveExternalFilesActivity.java [hunk 7]
```diff
@@ -437,7 +437,7 @@ private void populateDirectoryList() {
 
         mFile = getStorageManager().getFileByPath(full_path);
         if (mFile != null) {
-            Vector<OCFile> files = getStorageManager().getFolderContent(mFile, false);
+            Vector<OCFile> files = getStorageManager().getFolderContent(mFile, false,false);
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/ui/activity/ReceiveExternalFilesActivity.java]: Indent this file

File: owncloudApp/src/main/java/com/owncloud/android/ui/fragment/ExtendedListFragment.java [hunk 8]
```diff
@@ -481,4 +483,9 @@ boolean isShowingOnlyAvailableOffline() {
         Bundle args = getArguments();
         return ((args != null) && args.getBoolean(ARG_ONLY_AVAILABLE_OFFLINE, false));
     }
+
+    boolean isShowingSharedByLinkFiles(){
+        Bundle args = getArguments();
+        return ((args != null) && args.getBoolean(ARG_SHARED_BY_LINK_FILES,false));
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/ui/fragment/ExtendedListFragment.java]: Same

File: owncloudApp/src/main/java/com/owncloud/android/ui/adapter/FileListListAdapter.java [hunk 9]
```diff
@@ -392,8 +395,12 @@ public void swapDirectory(OCFile folder, FileDataStorageManager updatedStorageMa
             if (mOnlyAvailableOffline && (folder.equals(updatedStorageManager.getFileByPath(OCFile.ROOT_PATH)) ||
                     !folder.isAvailableOffline())) {
                 mImmutableFilesList = updatedStorageManager.getAvailableOfflineFilesFromCurrentAccount();
-            } else {
-                mImmutableFilesList = mStorageManager.getFolderContent(folder, mOnlyAvailableOffline);
+            } else if(mSharedByLinkFiles && (folder.equals(updatedStorageManager.getFileByPath(OCFile.ROOT_PATH)) ||
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/ui/adapter/FileListListAdapter.java]: Same

File: owncloudApp/src/main/java/com/owncloud/android/ui/fragment/FileDetailFragment.java [hunk 10]
```diff
@@ -239,7 +239,7 @@ public void onPrepareOptionsMenu(Menu menu) {
                     mContainerActivity,
                     getActivity()
             );
-            mf.filter(menu, false, false, false);
+            mf.filter(menu, false, false, false,false);
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/ui/fragment/FileDetailFragment.java]: Same

File: owncloudApp/src/main/java/com/owncloud/android/ui/fragment/OCFileListFragment.java [hunk 11]
```diff
@@ -141,14 +141,16 @@
...
         boolean hideFAB,
         boolean allowContextualMode
     ) {
         OCFileListFragment frag = new OCFileListFragment();
         Bundle args = new Bundle();
         args.putBoolean(ARG_JUST_FOLDERS, justFolders);
         args.putBoolean(ARG_ONLY_AVAILABLE_OFFLINE, onlyAvailableOffline);
-        if(onlyAvailableOffline) { hideFAB = true; }
+        args.putBoolean(ARG_SHARED_BY_LINK_FILES,sharedByLinkFiles);
+        if(onlyAvailableOffline || sharedByLinkFiles) { hideFAB = true; }
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/ui/fragment/OCFileListFragment.java]: Same

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.java [hunk 12]
```diff
@@ -318,7 +319,7 @@ public void onPrepareOptionsMenu(Menu menu) {
                 mContainerActivity,
                 getActivity()
         );
-        mf.filter(menu, false, false, false);
+        mf.filter(menu, false, false, false,false);
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewAudioFragment.java]: Same

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewImageFragment.java [hunk 13]
```diff
@@ -253,7 +254,7 @@ public void onPrepareOptionsMenu(Menu menu) {
                     mContainerActivity,
                     getActivity()
             );
-            mf.filter(menu, false, false, false);
+            mf.filter(menu, false, false, false,false);
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewImageFragment.java]: Same

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewTextFragment.java [hunk 14]
```diff
@@ -303,7 +304,7 @@ public void onPrepareOptionsMenu(Menu menu) {
                     mContainerActivity,
                     getActivity()
             );
-            mf.filter(menu, false, false, false);
+            mf.filter(menu, false, false,false, false);
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewTextFragment.java]: Same

File: owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoFragment.java [hunk 15]
```diff
@@ -350,7 +351,7 @@ public void onPrepareOptionsMenu(Menu menu) {
                 mContainerActivity,
                 getActivity()
         );
-        mf.filter(menu, false, false, false);
+        mf.filter(menu, false, false, false,false);
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewVideoFragment.java]: Same

**Comment by davigonz** [comment]
@abelgardep some changes

**Comment by davigonz** [comment]
👌 LGTM

**Comment by davigonz** [comment]
@jesmrec ready to test

**Comment by jesmrec** [comment]
### (2)

Does it make sense to have browsability in the "Shared by link" view? In the case of av. offline, it makes sense since the content inside an av. offline folder is also av. offline. However, a shared folder does not have to include shared content. That causes blank screens through browsing (correctly marked as "No content shared").

the question is... is it feasible not to browse through shared by link folders inside the view? 

Other approaches: iOS new app allows browsing inside the shared by link folder, displaying all the content inside, not only the shared one.

**Comment by davigonz** [comment]
After trying to fix [Folders browsability](https://github.com/owncloud/android/pull/2705#issuecomment-555408073) I do not fully understand this two scenarios in the PR code:

- Shared by link option in side menu: we were requesting shared by link files to the database when browsing shared parent folders. But we want to see all the files contained in a shared by link folder, even not the shared ones.
- Av offline option in side menu (not in this PR, but somehow related): we were requesting available offline files to the database when browsing av. offline parent folder. Not needed since all the available files contained in an available offline folder are always available offline.

Therefore, I've removed some extra parameters that are not needed in my opinion and now we are getting the folder content independently from we are in a shared or available offline folder. This way, browsability in shared folders through Shared by link shortcut work as expected.

@abelgardep since you rebased this contribution and had a look at the code, what's your opinion on it? Were those parameters needed for something I'm missing?

> the question is... is it feasible not to browse through shared by link folders inside the view?

I thought so but the server allows browsing inside the shared by link folder and iOS as well.

The last commit of this PR allows this browsing.

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt [hunk 16]
```diff
@@ -184,7 +185,7 @@ class DocumentsStorageProvider : DocumentsProvider() {
         val resultCursor = FileCursor(projection)
 
         // Create result cursor before syncing folder again, in order to enable faster loading
-        currentStorageManager?.getFolderContent(getFileById(folderId), false)
+        currentStorageManager?.getFolderContent(currentStorageManager?.getFileById(folderId))
```
Comment (abelgardep) [inline owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: I think that `currentStorageManager?.` is not needed, is it?

**Comment by abelgardep** [comment]
Sure, you are right. There is no need to filter every folder if we want to show every file/folder inside an available offline or shared by link. So i think that is ready to QA again @jesmrec

**Comment by davigonz** [comment]
@jesmrec https://github.com/owncloud/android/pull/2705#issuecomment-555400930 is ready to test as well

**Comment by jesmrec** [comment]
### (3)

1. Create an empty folder (0B)
2. Create a link over the folder
3. Go to "Shared by link" option in drawer menu
4. Sort by size

Current: 0B items are not correctly placed, as well as other stuff:

|  |  |
| :--: | :--:|
| device-2019-12-03-120813| device-2019-12-03-121053
|

Expected: correct sorting

Xiaomi MiA2 v9

**Comment by jesmrec** [comment]
### (4)

Just one clarification: list of links does not depend on the links created in the account, but the links locally discovered. That means, a link created in webUI or other client may not be listed in the Android client. In a new iteration, we can improve by getting al the shares from API:

`https://doc.owncloud.com/server/developer_manual/core/apis/ocs-share-api.html#get-all-shares`

this will have some concerns (what to do if you browse inside the item). Further ideas welcome.

**Comment by davigonz** [comment]
> Just one clarification: list of links does not depend on the links created in the account, but the links locally discovered. That means, a link created in webUI or other client may not be listed in the Android client.

Not exactly, Shared by links shortcut is just a filter applied over the files stored in database, getting the ones that have been shared, so the files that have been recently shared (from other clients) and have not been synchronized before (from all files view) will not appear there. 

> That means, a link created in webUI or other client may not be listed in the Android client.

A link created in webUI or other client will be listed in the Android client as long as the file synchronization has taken place and this only happens from the files view.

> https://doc.owncloud.com/server/developer_manual/core/apis/ocs-share-api.html#get-all-shares

In the Share by link options we are working with shared files, not with shares themselves. That API resource is used for getting all shares shared with a user.

**Comment by jesmrec** [comment]
that's what i meant, you see the content shared by link from the discovered items. If you did not browse yet, in the filtered list you will see only the "shared by link" content in the root folder.  

> In the Shared by link option we are working with shared files, not with shares themselves. That API resource is used for getting all shares shared with a user.

i know, this could be a way to get all the "shared by link" items without discovering them ( `share_type` = 3). It concerns also discovering such items in order to be able to browse through in the filtered list... more complicated.

So, let's get the current approach.

**Comment by jesmrec** [comment]
Let's move forward then. Approved

**Pull request merged**